File: spec/lib/data_structure/Heap_Spec.rb
require 'spec_helper'

describe Heap do

  describe 'when inserting' do
    its :front do
      subject.insert 5

      should eq 5
    end

    its :front do
      subject.insert 10

      should eq 10
    end

    its :front do
      inserts 2, 5, 10

      should eq 10
    end

    its :front do
      inserts 2, 3, 4

      should eq 4
    end

    its :front do
      inserts 4, 2, 6, 34, 7, 89, 5

      should eq 89
    end
  end

  describe 'when deleting' do
    its :front do
      inserts 5, 10
      subject.delete_front

      should eq 5
    end

    its :front do
      inserts 2, 5, 10
      subject.delete_front

      should eq 5
    end

    its :front do
      inserts 2, 10, 5, 15
      subject.delete_front

      should eq 10
    end

    its :front do
      inserts 2, 3, 4
      subject.delete_front

      should eq 3
    end

    its :front do
      inserts 4, 2, 6, 34, 7, 89, 5
      subject.delete_front

      should eq 34
    end
  end

  def inserts *args
    args.each do |arg|
      subject.insert arg
    end
  end
end

File: spec/javascripts/dataStructure/HeapSpec.js
describe('Heap', function() {

  xit('should insert a value and put the max value on the top', function() {
    heap = new Heap();

    heap.insert(5);
    expect(heap.front()).toEqual(5);

    heap.insert(4);
    expect(heap.front()).toEqual(5);

    heap.insert(10);
    expect(heap.front()).toEqual(10);
  });

  it('should delete the top value and replace it with the bottom then sort it', function() {
    heap = new Heap();

    inserts(heap, 1, 2, 3, 4, 5);
    expect(heap.front()).toEqual(5);
    heap.deleteFront();
    expect(heap.front()).toEqual(4);

    inserts(heap, 56, 34, 5);
    expect(heap.front()).toEqual(56);
    heap.deleteFront();
    expect(heap.front()).toEqual(34);
  });

  var inserts = function() {
    for(var i = 1; i < arguments.length; i++) {
      arguments[0].insert(arguments[i]);
    }
  };
});
